metagraph/src/graph/representation/canonical_dbg.cpp

```diff
@@ -62,79 +62,100 @@ void CanonicalDBG
 ::map_to_nodes_sequentially(std::string_view sequence,
                             const std::function<void(node_index)> &callback,
                             const std::function<bool()> &terminate) const {
-    std::vector<node_index> path = map_sequence_to_nodes(graph_, sequence);
-    auto first_not_found = std::find(path.begin(), path.end(), DeBruijnGraph::npos);
+    if (sequence.size() < get_k())
+        return;
+
+    std::vector<node_index> path;
+    path.reserve(sequence.size() - get_k() + 1);
+
+    // map until the first mismatch
+    bool stop = false;
+    graph_.map_to_nodes_sequentially(sequence,
+        [&](node_index node) {
+            if (node) {
+                path.push_back(node);
+            } else {
+                stop = true;
+            }
+        },
+        [&]() { return stop; }
+    );
 
-    for (auto jt = path.begin(); jt != first_not_found; ++jt) {
+    for (node_index node : path) {
         if (terminate())
             return;
 
-        callback(*jt);
+        callback(node);
     }
 
-    if (first_not_found == path.end())
+    // trim the mapped prefix
+    sequence = sequence.substr(path.size());
+    path.resize(0);
+    if (sequence.size() < get_k())
         return;
 
-    std::string rev_seq(sequence.begin() + (first_not_found - path.begin()),
-                        sequence.end());
+    std::string rev_seq(sequence);
     ::reverse_complement(rev_seq.begin(), rev_seq.end());
-    std::vector<node_index> rev_path;
-    if (const auto *dbg_succ = dynamic_cast<const DBGSuccinct*>(&graph_)) {
-        const auto &boss = dbg_succ->get_boss();
+    // map the reverse-complement
+    std::vector<node_index> rev_path = map_sequence_to_nodes(graph_, rev_seq);
```

Comment:
Wouldn't it be faster to again do a `graph_.traverse` or a `graph_.map_to_nodes_sequentially` call with rev_seq? then you can alternate between calling this on the forward and reverse complement orientation.

I'm not sure if doing things this way is that much faster, since you still end up doing the lookups and not finding k-mers on `rev_seq`.

Reply:
To alternate between them, we'd need to synchronously map the opposite ends of the sequences.
This means we'd have to map the rev-compl sequence backwards, which is a bit slower.
So I decided to do it forwards for now. Though in this case, we can't alternate, so I just map the entire thing at once.

One can try later to do it via backward mapping. But that will need benchmarking, and it's not clear if that's going to be faster than the version implemented here.

The current version tries to map from both ends.
Consider these 3 important cases:
* all k-mers in forward are found   => Fast mapping, no misses. Though the old version would do well too.
* all k-mers in forward are missing, but all in reverse are found   => Fast mapping, 1 miss. The old version would do n cold searches of k-mers in forward. So the new version is almost k times faster.
* more general, a prefix of forward and prefix of rev-compl are found    => similar to the previous case
* symmetric: a suffix of forward and suffix of rev-compl are found    => same as before, no speedup

Reply:
thinking about this again, if it's done in an alternating fashion, then it'll end up doing more full-k lookups...

Reply:
During annotation, most of the mapped contigs should belong to case 1 or case 2.
So, this optimization should make it on average about k/2 times faster.